src/performance/index.ts

```diff
@@ -0,0 +1,55 @@
+
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { CustomPerfOptionsType } from '../types';
+import Report from '../services/report';
+import pagePerf from './perf';
+
+class TracePerf {
+  private isPerf: boolean = true;
+  private perfConfig = {
+    resources: [],
+    perfDetail: {},
+  } as any;
+
+  public recordPerf(options: CustomPerfOptionsType) {
+    if (this.isPerf) {
+      this.perfConfig.performance = pagePerf.getPerfTiming();
+    }
+    const perfInfo = {
+      perfDetail: this.perfConfig.perfDetail,
+      resources: this.perfConfig.resources,
+      ...options,
+    };
+    new Report(options.reportUrl).sendByXhr(perfInfo);
+    this.clearPerf();
```

Comment:
What happens if the register hasn't completed?

Reply:
I don't understand your question. The register of perf is `tracePerfDetail`. It doesn't seem to have the problem.

Reply:
I set the question for a wrong line, sorry. I mean for `#sendByXhr`. According to the protocol, https://github.com/apache/skywalking-data-collect-protocol/blob/master/browser/BrowserPerf.proto#L39-L41, the report data should include service id and version id. But I can't see the check and buffer(if not registered yet).

Reply:
Oooh,  are these two parameters optional? If we don't pass ids, server seems to get data from service name and version id is current version. I forgot where it was explained. BTW, if we analysis data with time dimension, version doesn't seem to be need.

Reply:
No, those two parameters are required, that is why I asked. 
1. Version could be set as `-` by default, if the client-js doesn't set it, but this must notify the backend implementation.
1. Service Name is not existing in the protocol, only service ID, which means you have registered successfully. Do you want to add a name to the protocol as an alternative way?

My point is, this register would take time, you should have a buffer to hold this value until the register is done. Unless you should add name and version strings into the protocol. Also, you should know, the ID is required in the distributed tracing too. Even you add name and version in this protocol, you can't add those in the tracing protocol, the buffer mechanism is still required.

Reply:
I update the PR, and I think we need service names to show in analysis UI.